library/init/meta/widget/basic.lean

```diff
@@ -279,19 +279,30 @@ meta def with_cn : string → html α → html α
 meta def with_key {β} [has_to_string β] : β → html α → html α
 | s h := with_attr (key s) h
 
+/-- An effect is an action at the root of the widget component hierarchy
+and can give instructions to the editor to perform some task. -/
+meta inductive effect : Type
+| insert_text (text : string)
+| reveal_position (file_name : string) (p : pos)
+| highlight_position (file_name : string) (p : pos)
+| clear_highlighting
+| custom (key : string) (value : string)
```

Comment:
What do you think the `custom` effect will be used for?  You need to modify the vscode extension anyhow for a new effect type.

Reply:
Yeah it just means that you can test out new effects without having to have a custom build of lean was the idea.